### PR TITLE
unitary event analysis plot

### PR DESCRIPTION
**Pull request by dizcza**
Reopened https://github.com/INM-6/viziphant-new/pull/6

### CODE REVIEW

File: viziphant/unitary_event_analysis.py [hunk 1]
```diff
@@ -0,0 +1,391 @@
...
+            The y-axis limits for the joint surprise plot.
+        marker_size : integers (default: 5)
+            The marker size for the coincidence and unitary events.
+        y_tick_interval : integers (default: 15)
+            The size of the interval between y-ticks (for rasterplots only)
+        showfig : boolean (default: True)
+            Displays the figure on screen if True.
+        savefig : boolean (default: False)
+            Saves the figure to disk if True.
+        path_filname_format : string (default: figure.pdf)
```
Comment (Qkley) [inline viziphant/unitary_event_analysis.py:129]: Typo: `path_filename_format`.
```suggestion
        path_filename_format : string (default: figure.pdf)
```

**Comment by Qkley** [comment]
Viziphant Release 0 1
Viziphant-new PR

I also suggest you have a look at the comparison of the same analysis visualization using Viziphant 0.1 release versus your implementation. *Some unnecessary changes have been made:*
* The events horizontal line is missing from Spike Events plot
* The lw has been reduced to 0.5, making the figures unreadable in some contexts (lw was 2 before)
* The unit separator (horizontal line, in Events plots has been changed from black to purple)
* The negative alpha line in the significance threshold was changed from green to purple.
* The legends and unit names are on the right of the figure, making it unnecessarily large.
* The units of xaxis label Times should be in parentheses (), not square brackets [].

I suggest you also address these points by returning the aforementioned aspects to look closer to the original.

File: viziphant/tests/test_unitary_event_analysis.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        cls.significance_level = 0.05
+        cls.bin_size = 5 * pq.ms
+        cls.window_size = 100 * pq.ms
+        cls.window_step = 10 * pq.ms
+        cls.n_neurons = 2
+        cls.plot_params_user = {'events': {'Vision': [1000] * pq.ms,
+                                           'Action': [1500] * pq.ms},
+                                'savefig': True, 'showfig': False}
+
+    def _do_plot_UE(self, plot_path):
```
[the remark this answers is not in the document]
Comment (dizcza) [reply viziphant/tests/test_unitary_event_analysis.py:48]: make/do - does not matter. it's a helper for tests
plot_UE will be renamed

File: viziphant/tests/test_unitary_event_analysis.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+                                       'Action': [1500] * pq.ms},
+                            'savefig': True, 'showfig': False,
+                            'path_filename_format': plot_path}
+        plot_UE(self.spiketrains, joint_surprise_dict=self.UE,
+                significance_level=0.05, bin_size=5 * pq.ms,
+                window_size=100 * pq.ms, window_step=10 * pq.ms, n_neurons=2,
+                **plot_params_user)
+
+    def test_plot_UE(self):
+        # TODO: fix UE target plot once uploaded
```
[the remark this answers is not in the document]
Comment (dizcza) [reply viziphant/tests/test_unitary_event_analysis.py:59]: it's not a good practice unless we have a reason
we do have a reason for we didn't set up gin storage for viziphant target plot images
ignore these todos

File: viziphant/unitary_event_analysis.py [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+    # show figure
+    'showfig': True,
+    # save figure
+    'savefig': False,
+    # path, file name and format for saving the figure
+    'path_filname_format': 'figure.pdf'
+}
+
+
+def plot_UE(spiketrains, joint_surprise_dict, significance_level, bin_size,
```
[the remark this answers is not in the document]
Comment (dizcza) [reply viziphant/unitary_event_analysis.py:50]: rename it to `plot_unitary_events`

File: viziphant/unitary_event_analysis.py [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+    axis1.set_yticklabels(y_ticks_labels_list, fontsize=params_dict['fsize'])
+    for neuron_id in range(n_neurons):
+        axis1.text(xlim_right + 20, neuron_id * (n_trial + 1),
+                   f"Unit {params_dict['unit_real_ids'][neuron_id]}")
+    axis1.set_ylabel('Trial', fontsize=params_dict['fsize'])
+
+    print('plotting Spike Rates ...')
+    axis2 = plt.subplot(6, 1, 2, sharex=axis1)
+    axis2.set_title('Spike Rates')
+    # psth = peristimulu time histogram
```
[the remark this answers is not in the document]
Comment (dizcza) [reply viziphant/unitary_event_analysis.py:233]: why removing? I find the comment useful

File: viziphant/unitary_event_analysis.py [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+               color='b')
+    axis5.xaxis.set_major_locator(MaxNLocator(integer=True))
+    axis5.set_yticks([ue.jointJ(0.99), ue.jointJ(0.5), ue.jointJ(0.01)])
+    mark_epochs(axis5)
+    axis5.set_yticklabels([0.99, 0.5, 0.01])
+
+    print('plottting Unitary Events ...')
+    axis6 = plt.subplot(6, 1, 6, sharex=axis1)
+    axis6.set_title('Unitary Events')
+    # sig_inx_win: indices of significant JointSurprises within a window
```
[the remark this answers is not in the document]
Comment (dizcza) [reply viziphant/unitary_event_analysis.py:331]: good point

**Comment by dizcza** [comment]
> The events horizontal line is missing from Spike Events plot
The lw has been reduced to 0.5, making the figures unreadable in some contexts (lw was 2 before)
The unit separator (horizontal line, in Events plots has been changed from black to purple)
The negative alpha line in the significance threshold was changed from green to purple.
The legends and unit names are on the right of the figure, making it unnecessarily large.
The units of xaxis label Times should be in parentheses (), not square brackets [].

I agree with all (or rather don't have a preference for some of them). It's not hard to address them, so Max please do it.

File: viziphant/unitary_event_analysis.py [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+               label='expected', lw=params_dict['lw'], color='m')
+    axis4.set_xlim(xlim_left, xlim_right)
+    axis4.xaxis.set_major_locator(MaxNLocator(integer=True))
+    y_ticks = axis4.get_ylim()
+    axis4.set_yticks([0, y_ticks[1] / 2, y_ticks[1]])
+    mark_epochs(axis4)
+    axis4.legend(bbox_to_anchor=(1, 1), fancybox=True, shadow=True)
+    axis4.set_ylabel('(1/s)', fontsize=params_dict['fsize'])
+
+    print('plotting Statistical Significance ...')
```
[the remark this answers is not in the document]
Comment (ojoenlanuca) [reply viziphant/unitary_event_analysis.py:308]: so just the top-value changes?

File: viziphant/unitary_event_analysis.py [hunk 8]
```diff
@@ -0,0 +1,404 @@
...
+    axis5.set_yticklabels([1-significance_level, 0.5, significance_level])
+
+    print('plotting Unitary Events ...')
+    axis6 = plt.subplot(6, 1, 6, sharex=axis1)
+    axis6.set_title('Unitary Events')
+    for n in range(n_neurons):
+        for trial, data_trial in enumerate(data):
+            spike_events_on_timescale = data_trial[n].magnitude
+            spike_events_on_trialscale = \
+                np.ones_like(data_trial[n].magnitude) * trial + \
```
Comment (dizcza) [inline viziphant/unitary_event_analysis.py:347]: I see this thing for the 3rd time
hm
but you already spent a lot of time restructuring the logic. let it be as it is

File: viziphant/unitary_event_analysis.py [hunk 9]
```diff
@@ -0,0 +1,404 @@
...
+        Marks epochs on the respective axis by creating a vertical line and
+        shows the epoch's name under the last subplot. Epochs need to be
+        defined in the plot_params_user dictionary.
+        Parameters
+        ----------
+        axes_name : matplotlib.axes._subplots.AxesSubplot
+            The axis on which the epochs will be marked.
+        """
+        for key in params_dict['events'].keys():
+            for e_val in params_dict['events'][key]:
```
[the remark this answers is not in the document]
Comment (ojoenlanuca) [reply viziphant/unitary_event_analysis.py:215]: now e_val is: value_of_key

[the remark this answers is not in the document]
Comment (dizcza) [reply viziphant/unitary_event_analysis.py:215]: oh, that's even worse!
it bears the same information is `value` word and yet you made it longer by adding `_of_key` suffix that does not any information to the reader at all!
moreover, both `value` and `value_of_key` tell the user no more than any bad naming like `x` or `result`
think more

[the remark this answers is not in the document]
Comment (ojoenlanuca) [reply viziphant/unitary_event_analysis.py:215]: how is "e_timepoint" ?

[the remark this answers is not in the document]
Comment (dizcza) [reply viziphant/unitary_event_analysis.py:215]: why not event_timepoint?
my original comment was that `e` is not readable
but I see you keep it persistent, assuming that is clear that `e` == `event`. well, it's not.

[the remark this answers is not in the document]
Comment (ojoenlanuca) [reply viziphant/unitary_event_analysis.py:215]: ok than 'event_timepoint'

File: viziphant/unitary_event_analysis.py [hunk 10]
```diff
@@ -0,0 +1,401 @@
...
+                     statistical_significance_axis, unitary_events_axis
+    """
+    # update params_dict_default with user input
+    params_dict = params_dict_default.copy()
+    params_dict.update(plot_params_user)
+
+    # rescale all spiketrains to the uniform time unit from params_dict
+    for m in range(len(data)):
+        for n in range(len(data[0])):
+            data[m][n].rescale(params_dict['time_unit'])
```
Comment (Qkley) [inline viziphant/unitary_event_analysis.py:150]: Rescale returns a copy of the data; it is not an in place operation. In other words, this line does nothing. You either need to save the new copy or immediately do an operation with the returned copy. 
See here: https://github.com/python-quantities/python-quantities/blob/master/quantities/quantity.py (lines 203-207)

File: viziphant/unitary_event_analysis.py [hunk 11]
```diff
@@ -0,0 +1,401 @@
...
+                axes_name.axvline(event_timepoint, ls='-',
+                                  lw=params_dict['lw'], color='r')
+                if axes_name.get_geometry()[2] == 6:
+                    axes_name.text(x=event_timepoint-10*pq.ms, y=-90, s=key,
+                                   fontsize=12, color='b')
+
+    print('plotting Unitary Event Analysis ...')
+
+    print('plotting Spike Events ...')
+    axis1 = plt.subplot(6, 1, 1)
```
Comment (Qkley) [inline viziphant/unitary_event_analysis.py:221]: It should be `axes` not `axis`. Axes contain figure elements. Axis contains ticks, labels, etc.. Please change to `ax` or `axes` throughout.
See here: https://matplotlib.org/1.5.1/faq/usage_faq.html#parts-of-a-figure

File: viziphant/unitary_event_analysis.py [hunk 12]
```diff
@@ -0,0 +1,401 @@
...
+
+import matplotlib.pyplot as plt
+import numpy as np
+import quantities as pq
+from matplotlib.ticker import (MaxNLocator)
+
+import elephant.unitary_event_analysis as ue
+
+FigureUE = namedtuple(
+    "AxesUE", "spike_events, spike_rates, coincidence_events, "
```
Comment (Qkley) [inline viziphant/unitary_event_analysis.py:16]:
```suggestion
    "AxesUE", "spike_events, spike_rates, coincident_events, "
```

File: viziphant/unitary_event_analysis.py [hunk 13]
```diff
@@ -0,0 +1,401 @@
...
+    axis6.set_yticks(y_ticks_list)
+    axis6.set_yticklabels(y_ticks_labels_list, fontsize=params_dict['fsize'])
+    axis6.set_xlabel('Time (ms)', fontsize=params_dict['fsize'])
+    axis6.set_ylabel('Trial', fontsize=params_dict['fsize'])
+
+    # mark all epochs on all subplots and annotate all subplotaxis
+    for n in range(1, 7):
+        mark_epochs(eval(f"axis{n}"))
+        axis = locals()['axis'+str(n)]
+        axis.text(-0.05, 1.1, string.ascii_uppercase[n-1],
```
Comment (Qkley) [inline viziphant/unitary_event_analysis.py:396]: Explicit is better than implicit. Consider explicitly adding the letters next to plots.

**Comment by Qkley** [comment]
I have not tested the script , but I think it is necessary to do so since the script has extensive changes compared to the previous version.

File: viziphant/unitary_event_analysis.py [hunk 14]
```diff
@@ -0,0 +1,401 @@
...
+        axes_name : matplotlib.axes._subplots.AxesSubplot
+            The axis on which the epochs will be marked.
+        """
+        for key in params_dict['events'].keys():
+            for event_timepoint in params_dict['events'][key]:
+                axes_name.axvline(event_timepoint, ls='-',
+                                  lw=params_dict['lw'], color='r')
+                if axes_name.get_geometry()[2] == 6:
+                    axes_name.text(x=event_timepoint-10*pq.ms, y=-90, s=key,
+                                   fontsize=12, color='b')
```
Comment (Qkley) [inline viziphant/unitary_event_analysis.py:216]: This function needs testing, in my opinion.

File: viziphant/unitary_event_analysis.py (hunk 13 above)
Comment (ojoenlanuca) [reply viziphant/unitary_event_analysis.py:396]: do you mean like: (e.g)
axes1.text(-0.05, 1.1, 'A',
                  transform=axes.transAxes, size=params_dict['fsize'] + 5,
                  weight='bold')

File: viziphant/unitary_event_analysis.py (hunk 14 above)
Comment (dizcza) [reply viziphant/unitary_event_analysis.py:216]: @Qkley how would you proceed with testing this internal function? Do you mean testing by playing with parameters and asserting that it's OK just by looking at it or you mean unittests?
In the latter case, it should be moved outside of the main function - plot_unitary_events - which I don't find reasonable.

Comment (Qkley) [reply viziphant/unitary_event_analysis.py:216]: I mean functional testing, not unit tests. I will load the feature/unitary_event_analysis as my viziphant package on Pycharm and try to generate plots using real data. I will pass some boundary values to test how `mark_epochs` function handles them. Think of it as black-box testing to ensure that the unitary_event_analysis does what it is supposed to do.

File: viziphant/unitary_event_analysis.py (hunk 13 above)
Comment (Qkley) [reply viziphant/unitary_event_analysis.py:396]: Something like this:
```
axes_list = [axes1, axes2, axes3, axes4, axes5, axes6]
letter_list = ['A', 'B', 'C', 'D', 'E', 'F']
for (axes, letter) in zip(axes_list, letter_list):
    axes.text(-0.05, 1.1, letter, transform=axes.transAxes, size=params_dict['fsize'] + 5,
                    weight='bold')
```
However, please do not just copy-paste the lines above as it is untested. Also, Danylo, feel free to comment on the approach.

**Comment by Qkley** [comment]
_Vizphant 0.1 Release_
Viziphant Release 0 1

_This PR_:
Viziphant-new_PR_test_27MAY20_1810160011_1810160025_tactile_correct1_analytic_TrialByTrial
* Unit name is hidden behind a panel B
* Panel C title should be **Coincident Events**
* You should keep legends uniform (i.e. capitalize **Empirical** and **Expected**)
* Nice work with the Spike Rates plot!

_Events test_ (passed 4 events at -100, 300, 1000 and 2200 ms)
Viziphant-new_PR_events_test_27MAY20_1810160011_1810160025_tactile_correct1_analytic_TrialByTrial
The events are missing their labels on the plot (see Viziphant 0.1 Release plot above)

_Significance test_ (changed significance to 0.01)
Viziphant-new_PR_significance_test_27MAY20_1810160011_1810160025_tactile_correct1_analytic_TrialByTrial
Nice work dealing with 0.01 and 0.05 significance!

**Comment by Qkley** [comment]
I've tested the updated code without any params except for the event times and unit ids.
Viziphant-new_PR_events_test_28MAY20_1810160011_1810160025_tactile_correct1_analytic_TrialByTrial

It looks to me that you have changed some **default values** for plotting parameters.
* Figsize was (10, 12) and now is (12, 10).
* You have also changed figure margins
See Viziphant 0.1 Release default params here: https://github.com/INM-6/viziphant/blob/master/viziphant/unitary_event_analysis.py

This might be the reason why old plots do not look like new plots.

When it comes to the new commits:
* Labels are plotted for the events outside of the figure limit (Min bound is at - 100 pq*ms)
* Also label DM overlaps with the xaxis label Time (ms)

File: viziphant/unitary_event_analysis.py [hunk 15]
```diff
@@ -209,12 +186,15 @@ def mark_epochs(axes_name):
...
-                                   fontsize=12, color='b')
+                # check if epochs are between time-axis limits
+                if ((xlim_left <= event_timepoint) and
+                        (event_timepoint <= xlim_right)):
+                    axes_name.axvline(event_timepoint, ls='-',
+                                      lw=params_dict['lw'], color='r')
+                    if axes_name.get_geometry()[2] == 6:
+                        axes_name.text(x=event_timepoint-10*pq.ms,
+                                       y=-54, s=key,
+                                       fontsize=12, color='r')
```
Comment (Qkley) [inline viziphant/unitary_event_analysis.py:197]: You do not have to hardcode an offset for the event labels. Matplotlib has options for this:
```suggestion
                        axes_name.text(x=event_timepoint,
                                       y=-54, s=key,
                                       fontsize=12, color='r',
                                       horizontalalignment='center')
```
`horizontalalignment` takes values `'left'`, `'center'`, `'right'`.  A demo can be found [here](https://matplotlib.org/3.1.1/gallery/text_labels_and_annotations/text_alignment.html#sphx-glr-gallery-text-labels-and-annotations-text-alignment-py).
